File: dart/lib/scheme/scheme_fidelity.dart
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import 'dart:math' as math;

import 'package:material_color_utilities/dislike/dislike_analyzer.dart';
import 'package:material_color_utilities/hct/hct.dart';
import 'package:material_color_utilities/palettes/tonal_palette.dart';
import 'package:material_color_utilities/temperature/temperature_cache.dart';
import 'dynamic_scheme.dart';
import 'variant.dart';

/// A scheme that places the source color in [Scheme.primaryContainer].
/// 
/// Primary Container is the source color, adjusted for color relativity.
/// It maintains constant appearance in light mode and dark mode. 
/// This adds ~5 tone in light mode, and subtracts ~5 tone in dark mode. 
///
/// Tertiary Container is the complement to the source color, using
/// [TemperatureCache]. It also maintains constant appearance.
class SchemeFidelity extends DynamicScheme {
  SchemeFidelity({
    required Hct sourceColorHct,
    required super.isDark,
    required super.contrastLevel,
  }) : super(
          sourceColorArgb: sourceColorHct.toInt(),
          variant: Variant.fidelity,
          primaryPalette: TonalPalette.of(
            sourceColorHct.hue,
            sourceColorHct.chroma,
          ),
          secondaryPalette: TonalPalette.of(
            sourceColorHct.hue,
            math.max(sourceColorHct.chroma - 32.0, sourceColorHct.chroma * 0.5),
          ),
          tertiaryPalette: TonalPalette.fromHct(
            DislikeAnalyzer.fixIfDisliked(
              TemperatureCache(sourceColorHct).complement,
            ),
          ),
          neutralPalette: TonalPalette.of(
            sourceColorHct.hue,
            sourceColorHct.chroma / 8.0,
          ),
          neutralVariantPalette: TonalPalette.of(
            sourceColorHct.hue,
            (sourceColorHct.chroma / 8.0) + 4.0,
          ),
        );
}

File: cpp/scheme/scheme_fidelity.cc
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/scheme/scheme_fidelity.h"

#include "cpp/cam/hct.h"
#include "cpp/dislike/dislike.h"
#include "cpp/dynamiccolor/dynamic_scheme.h"
#include "cpp/dynamiccolor/variant.h"
#include "cpp/palettes/tones.h"
#include "cpp/temperature/temperature_cache.h"

namespace material_color_utilities {

SchemeFidelity::SchemeFidelity(Hct set_source_color_hct, bool set_is_dark,
                               double set_contrast_level)
    : DynamicScheme(
          /*source_color_argb:*/ set_source_color_hct.ToInt(),
          /*variant:*/ Variant::kFidelity,
          /*contrast_level:*/ set_contrast_level,
          /*is_dark:*/ set_is_dark,
          /*primary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(),
                       set_source_color_hct.get_chroma()),
          /*secondary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(),
                       fmax(set_source_color_hct.get_chroma() - 32.0,
                            set_source_color_hct.get_chroma() * 0.5)),
          /*tertiary_palette:*/
          TonalPalette(FixIfDisliked(
              TemperatureCache(set_source_color_hct).GetComplement())),
          /*neutral_palette:*/
          TonalPalette(set_source_color_hct.get_hue(),
                       set_source_color_hct.get_chroma() / 8.0),
          /*neutral_variant_palette:*/
          TonalPalette(set_source_color_hct.get_hue(),
                       set_source_color_hct.get_chroma() / 8.0 + 4.0)) {}

SchemeFidelity::SchemeFidelity(Hct set_source_color_hct, bool set_is_dark)
    : SchemeFidelity::SchemeFidelity(set_source_color_hct, set_is_dark, 0.0) {}

}  // namespace material_color_utilities
